### ISSUE TITLE
Script apparently succeeds to run, but no blob in .tmp

### ISSUE THREAD

**Issue by hipp013** [Merculous/SHSHDump#1]
Running on fresh Ubuntu install, manually compiled libgeneral and img4tool successfully. Not sure if I had to do that for this script to work properly, but regardless, the script appears to succeed, however the blob is not generated. 

```
andrew@andrew-MacBookPro:~$ export PKG_CONFIG_PATH="/usr/local/lib/pkgconfig"

andrew@andrew-MacBookPro:~$ git clone https://github.com/merculous/SHSHDump

Cloning into 'SHSHDump'...
remote: Enumerating objects: 13, done.
remote: Counting objects: 100% (13/13), done.
remote: Compressing objects: 100% (8/8), done.
remote: Total 13 (delta 4), reused 13 (delta 4), pack-reused 0
Unpacking objects: 100% (13/13), 4.51 KiB | 769.00 KiB/s, done.

andrew@andrew-MacBookPro:~$ cd SHSHDump/
andrew@andrew-MacBookPro:~/SHSHDump$ python3 -m venv venv
andrew@andrew-MacBookPro:~/SHSHDump$ . venv/bin/activate
(venv) andrew@andrew-MacBookPro:~/SHSHDump$ pip install wheel

Collecting wheel
  Using cached wheel-0.35.1-py2.py3-none-any.whl (33 kB)
Installing collected packages: wheel
Successfully installed wheel-0.35.1

(venv) andrew@andrew-MacBookPro:~/SHSHDump$ pip install -r requirements.txt

Processing /home/andrew/.cache/pip/wheels/23/5c/73/e8ff4b159f2488175972bd4b8f0bb075318c634ea90237e987/autopep8-1.5.3-py2.py3-none-any.whl
Collecting bcrypt==3.1.7
  Using cached bcrypt-3.1.7-cp34-abi3-manylinux1_x86_64.whl (56 kB)
Collecting cffi==1.14.0
  Using cached cffi-1.14.0-cp38-cp38-manylinux1_x86_64.whl (409 kB)
Collecting cryptography==2.9.2
  Using cached cryptography-2.9.2-cp35-abi3-manylinux2010_x86_64.whl (2.7 MB)
Collecting flake8==3.8.3
  Using cached flake8-3.8.3-py2.py3-none-any.whl (72 kB)
Collecting importlib-metadata==1.7.0
  Using cached importlib_metadata-1.7.0-py2.py3-none-any.whl (31 kB)
Collecting mccabe==0.6.1
  Using cached mccabe-0.6.1-py2.py3-none-any.whl (8.6 kB)
Collecting paramiko==2.7.1
  Using cached paramiko-2.7.1-py2.py3-none-any.whl (206 kB)
Collecting pycodestyle==2.6.0
  Using cached pycodestyle-2.6.0-py2.py3-none-any.whl (41 kB)
Collecting pycparser==2.20
  Using cached pycparser-2.20-py2.py3-none-any.whl (112 kB)
Collecting pyflakes==2.2.0
  Using cached pyflakes-2.2.0-py2.py3-none-any.whl (66 kB)
Collecting PyNaCl==1.4.0
  Using cached PyNaCl-1.4.0-cp35-abi3-manylinux1_x86_64.whl (961 kB)
Collecting six==1.15.0
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting toml==0.10.1
  Using cached toml-0.10.1-py2.py3-none-any.whl (19 kB)
Collecting zipp==3.1.0
  Using cached zipp-3.1.0-py3-none-any.whl (4.9 kB)
Installing collected packages: toml, pycodestyle, autopep8, six, pycparser, cffi, bcrypt, cryptography, pyflakes, mccabe, flake8, zipp, importlib-metadata, PyNaCl, paramiko
Successfully installed PyNaCl-1.4.0 autopep8-1.5.3 bcrypt-3.1.7 cffi-1.14.0 cryptography-2.9.2 flake8-3.8.3 importlib-metadata-1.7.0 mccabe-0.6.1 paramiko-2.7.1 pycodestyle-2.6.0 pycparser-2.20 pyflakes-2.2.0 six-1.15.0 toml-0.10.1 zipp-3.1.0

(venv) andrew@andrew-MacBookPro:~/SHSHDump$ ./dump.py --ip localhost --port 2222 --user root --password alpine

Ensuring needed dependencies are installed!
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Note, selecting 'python-dev-is-python2' instead of 'python-dev'
automake is already the newest version (1:1.16.1-4ubuntu6).
cmake is already the newest version (3.16.3-1ubuntu1).
g++ is already the newest version (4:9.3.0-1ubuntu2).
libtool is already the newest version (2.4.6-14).
openssl is already the newest version (1.1.1f-1ubuntu2).
pkg-config is already the newest version (0.29.1-0ubuntu4).
python-dev-is-python2 is already the newest version (2.7.17-4).
0 upgraded, 0 newly installed, 0 to remove and 131 not upgraded.
Cloning into '.tmp/libplist'...
remote: Enumerating objects: 40, done.
remote: Counting objects: 100% (40/40), done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 4344 (delta 13), reused 21 (delta 7), pack-reused 4304
Receiving objects: 100% (4344/4344), 1.29 MiB | 3.51 MiB/s, done.
Resolving deltas: 100% (2679/2679), done.
configure.ac:29: installing './compile'
configure.ac:42: installing './config.guess'
configure.ac:42: installing './config.sub'
configure.ac:6: installing './install-sh'
configure.ac:6: installing './missing'
cython/Makefile.am: installing './depcomp'
parallel-tests: installing './test-driver'
ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')
ar: `u' modifier ignored since `D' is the default (see `U')
libtool: warning: relinking 'libplist++-2.0.la'
Cloning into '.tmp/libgeneral'...
remote: Enumerating objects: 20, done.
remote: Counting objects: 100% (20/20), done.
remote: Compressing objects: 100% (18/18), done.
remote: Total 162 (delta 6), reused 16 (delta 2), pack-reused 142
Receiving objects: 100% (162/162), 23.33 KiB | 2.33 MiB/s, done.
Resolving deltas: 100% (41/41), done.
aclocal: warning: couldn't open directory 'm4': No such file or directory
configure.ac:6: installing './compile'
configure.ac:4: installing './config.guess'
configure.ac:4: installing './config.sub'
configure.ac:8: installing './install-sh'
configure.ac:6: error: required file './ltmain.sh' not found
configure.ac:8: installing './missing'
libgeneral/Makefile.am: installing './depcomp'
aclocal: warning: couldn't open directory 'm4': No such file or directory
ar: `u' modifier ignored since `D' is the default (see `U')
Cloning into '.tmp/lzfse'...
remote: Enumerating objects: 203, done.
remote: Total 203 (delta 0), reused 0 (delta 0), pack-reused 203
Receiving objects: 100% (203/203), 98.55 KiB | 1.37 MiB/s, done.
Resolving deltas: 100% (116/116), done.
CMake Warning (dev) at CMakeLists.txt:60 (add_library):
  Policy CMP0069 is not set: INTERPROCEDURAL_OPTIMIZATION is enforced when
  enabled.  Run "cmake --help-policy CMP0069" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  INTERPROCEDURAL_OPTIMIZATION property will be ignored for target 'lzfse'.
This warning is for project developers.  Use -Wno-dev to suppress it.

Cloning into '.tmp/img4tool'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 1005 (delta 0), reused 1 (delta 0), pack-reused 1002
Receiving objects: 100% (1005/1005), 259.30 KiB | 2.29 MiB/s, done.
Resolving deltas: 100% (555/555), done.
aclocal: warning: couldn't open directory 'm4': No such file or directory
configure.ac:6: installing './compile'
configure.ac:4: installing './config.guess'
configure.ac:4: installing './config.sub'
configure.ac:8: installing './install-sh'
configure.ac:6: error: required file './ltmain.sh' not found
configure.ac:8: installing './missing'
img4tool/Makefile.am: installing './depcomp'
aclocal: warning: couldn't open directory 'm4': No such file or directory
img4tool.cpp:37:2: warning: #warning TODO adjust this for HAVE_COMMCRYPTO [-Wcpp]
 #warning TODO adjust this for HAVE_COMMCRYPTO
  ^
img4tool.cpp:777:22: warning: multi-character character constant [-Wmultichar]
                 case 'TSGD': //DGST
                      ^
img4tool.cpp:805:51: warning: multi-character character constant [-Wmultichar]
         ASN1DERElement bnch = getValFromIM4M(im4m,'BNCH');
                                                   ^
img4tool.cpp:1003:40: warning: multi-character character constant [-Wmultichar]
         assure(ptagVal == htonl('BNCN'));
                                        ^
img4tool.cpp:1004:77: warning: multi-character character constant [-Wmultichar]
         assure(*(uint32_t*)ptag[0].getStringValue().c_str() == htonl('BNCN'));
                                                                             ^
img4tool.cpp:1123:22: warning: multi-character character constant [-Wmultichar]
                 case 'DROB': //BORD
                      ^
img4tool.cpp:1127:22: warning: multi-character character constant [-Wmultichar]
                 case 'PIHC': //CHIP
                      ^
img4tool.cpp:1131:22: warning: multi-character character constant [-Wmultichar]
                 case 'MODS': //SDOM
                      ^
img4tool.cpp:1243:30: warning: multi-character character constant [-Wmultichar]
                         case 'TSGD': //DGST
                              ^
ar: `u' modifier ignored since `D' is the default (see `U')
Downloading ldid!
Extracting Linux ldid...
Starting dumping process!
Seems like /tmp/shsh_dump.bin exists!
Seems like everything worked! Enjoy your dumped shsh :P -Merc
We took 52.36 seconds!

(venv) andrew@andrew-MacBookPro:~/SHSHDump$ cd .tmp
(venv) andrew@andrew-MacBookPro:~/SHSHDump/.tmp$ ls -l

total 8456
drwxrwxr-x  9 andrew andrew    4096 Sep 10 01:28 img4tool
-rwxr-xr-x  1 andrew andrew   43480 Sep 10 01:28 ldid
-rw-rw-r--  1 andrew andrew  206559 Sep 10 01:28 ldid.zip
drwxrwxr-x  9 andrew andrew    4096 Sep 10 01:28 libgeneral
drwxrwxr-x 14 andrew andrew    4096 Sep 10 01:27 libplist
drwxrwxr-x  7 andrew andrew    4096 Sep 10 01:28 lzfse
-rw-rw-r--  1 andrew andrew 8388608 Sep 10 01:28 shsh_dump.bin
```

I recognize the errors between `img4tool.cpp:777` and `img4tool.cpp:1243` as being similar to the ones in your video demonstration, and the blob appears to dump successfully in your video, so I've so far ignored those errors.

If you remember helping me out on Twitter two weeks ago, I was running it on macOS at the time and running into different errors related to openssl. I have since installed an Ubuntu dualboot onto my Mac and ran the script from there. No openssl error, and everything has been compiled successfully, however no blob. 

The only thing I can think of is that I have to specify PKG_CONFIG_PATH as /usr/local/lib/pkgconfig in order for img4tool to find libgeneral. If I don't do that, it can't find libgeneral and will fail at that point. 

In my inexperienced opinion, I think the problem lies not with your script but with img4tool or libgeneral. However I was hoping you could help me pinpoint why the blob won't generate.

**Comment by Merculous** [comment]
Well, seems like everything worked, but img4tool. So, when you run img4tool, do you get any lib errors? Does is print out like normal or you getting an error like "unable to find lib x.so" or similar? Weird how it prints successful since I only have it print that if it actually was. Maybe I need to add better checking :/

`try:
        os.path.exists('.tmp/blob.shsh2')
    except FileNotFoundError:
        print('Something went wrong extracting the shsh/ticket from the dump!')
        raise
    else:
        print('Seems like everything worked! Enjoy your dumped shsh :P -Merc')`

This should only be printed if blob.shsh2 is actually present.You sure its not anywhere? Try using ls -la to show hidden files.

**Comment by hipp013** [comment]
It's not pictured in the above log, but `ls -la` produces the same result: no blob. 

Let me compile libgeneral and img4tool again. libgeneral compiles successfully, however img4tool does not unless I specify PKG_CONFIG_PATH. Once I run through the whole thing again, I'll edit this comment with the log.

**Comment by hipp013** [comment]
https://pastebin.com/95JieFsW

This time I'm running `ls -la` at the end to show it isn't there. I spaced out each section of the log for ease of reading, but let me clarify this is all the same log from one Terminal environment, with each step executed in the same order as it appears in the log.

Is there any way to run SHSHDump or any of its dependencies in a verbose mode or something, to get more information about what's going wrong?

**Comment by hipp013** [comment]
UPDATE: The problem is with img4tool. `-s`, `-v` and `--convert` are unavailable. This is the same issue I ran into on macOS so it is definitely not anything wrong with your script, but rather with img4tool.

```
(venv) andrew@andrew-MacBookPro:~$ img4tool
img4tool version: 0.182-22a2671082a9232c180e1856dfce04eea9681315
Compiled with plist: NO
Usage: img4tool [OPTIONS] FILE
Parses img4, im4p, im4m files

  -h, --help			prints usage information
  -a, --print-all		print everything from im4m
  -i, --im4p-only		print only im4p
  -e, --extract			extracts im4m/im4p payload
UNAVAILABLE:   -s, --shsh	<PATH>		Filepath for shsh (for reading/writing im4m)
  -m, --im4m	<PATH>		Filepath for im4m (depending on -e being set)
  -p, --im4p	<PATH>		Filepath for im4p (depending on -e being set)
  -c, --create	<PATH>		creates an img4 with the specified im4m, im4p or creates im4p with raw file (last argument)
  -o, --outfile			output path for extracting im4p payload (-e) or renaming im4p (-n)
  -t, --type			set type for creating IM4P files from raw
  -d, --desc			set desc for creating IM4P files from raw
  -n, --rename-payload NAME	rename im4p payload (NAME must be exactly 4 bytes)
UNAVAILABLE:   -v, --verify BUILDMANIFEST	verify img4, im4m
      --iv			IV  for decrypting payload when extracting (requires -e and -o)
      --key			Key for decrypting payload when extracting (requires -e and -o)
UNAVAILABLE:       --convert			convert IM4M file to .shsh (use with -s)
```

So if you were to make any changes on your end, maybe add error handling for `-s` `-v` and `--convert` being unavailable from img4tool.

**Comment by Merculous** [comment]
Weird. My script is automated to make everything needed. Libplist should've been made and that's what your img4tool is lacking. If libplist is missing in /usr/local/lib then recompile libplist and img4tool. Once, img4tool says it's compiled with PLIST, everything should work after that. I have my code to just convert the dump if it's already there.

**Comment by Merculous** [comment]
Tbh, this is more of tihmstars fault. I don't know why he compiles without plist support. It should just be default.

**Comment by hipp013** [comment]
Yeah at first I installed libplist via `brew install`, then I cloned it from Github and compiled it manually. Still didn't work. Even though your script ensures libplist is installed, would it be worth adding some sort of error handling to ensure those three options are available? 

I'm writing a comprehensive guide on how to compile everything yourself and run this script, so I'm trying to catch every single thing that can go wrong with img4tool in the process (and trust me, I've been documenting a *lot* of things that have been going wrong).
